kyo-core/shared/src/main/scala/kyo/env.scala

```diff
@@ -0,0 +1,47 @@
+package kyo
+
+import kyo.Tag.Intersection
+import scala.collection.immutable.HashMap
+
+opaque type Env[+R] = HashMap[Tag[?], Any]
+extension [R](self: Env[R])
+
+    private def fatal(using t: Tag[?]): Nothing =
+        throw new RuntimeException(s"fatal: kyo.Env of contents [$self] missing value of type: [${t.show}].")
+
+    def get[A >: R](using Tag[A]): A =
+        getOrElse(fatal)
+
+    def getOrElse[A >: R](default: => A)(using t: Tag[A]): A =
+        val result = self.getOrElse(t, null)
+        if isNull(result) then
+            default
+        else
+            result.asInstanceOf[A]
+        end if
+    end getOrElse
+
+    def add[A: Tag](a: A): Env[R & A] =
+        self.updated(Tag[A], a)
+
+    def union[R0](that: Env[R0]): Env[R & R0] =
```

Comment:
What do you think about naming this method `add` as well?

Reply:
The problem is that this not simply additive in the case of overlapping values. I think that may confuse users.